common/cache/budget.go

```diff
@@ -0,0 +1,1204 @@
+		return err
+	}
+	m.ReleaseBytesForCache(cacheID, nBytes)
+	return nil
+}
+
+func (m *manager) ReleaseCountWithCallback(cacheID string, nCount int64, callback func() error) error {
+	if err := callback(); err != nil {
+		return err
+	}
+	m.ReleaseCountForCache(cacheID, nCount)
+	return nil
+}
+
+func (m *manager) ReserveOrReclaimSelfReleaseWithCallback(ctx context.Context, cacheID string, nBytes uint64, nCount int64, retriable bool, reclaim ReclaimSelfRelease, callback func() error) error {
+	if err := m.ReserveOrReclaimSelfRelease(ctx, cacheID, nBytes, nCount, retriable, reclaim); err != nil {
+		return err
+	}
+	if err := callback(); err != nil {
+		m.ReleaseForCache(cacheID, nBytes, nCount)
+		return err
+	}
+	return nil
+}
+
+func (m *manager) ReserveOrReclaimManagerReleaseWithCallback(ctx context.Context, cacheID string, nBytes uint64, nCount int64, retriable bool, reclaim ReclaimManagerRelease, callback func() error) error {
+	if err := m.ReserveOrReclaimManagerRelease(ctx, cacheID, nBytes, nCount, retriable, reclaim); err != nil {
+		return err
+	}
+	if err := callback(); err != nil {
+		m.ReleaseForCache(cacheID, nBytes, nCount)
+		return err
+	}
+	return nil
+}
+
+func (m *manager) UsedBytes() uint64 { return atomic.LoadUint64(&m.usedBytes) }
+func (m *manager) UsedCount() int64  { return atomic.LoadInt64(&m.usedCount) }
+
+func (m *manager) CapacityBytes() uint64 {
+	v := m.maxBytes()
+	if v < 0 {
+		return math.MaxUint64
+	}
+	return uint64(v) // includes 0 as a valid, enforced cap
+}
+
+func (m *manager) CapacityCount() int64 {
+	v := m.maxCount()
+	if v < 0 {
+		return math.MaxInt64
+	}
+	// clamp just in case
+	if v > math.MaxInt64 {
+		return math.MaxInt64
+	}
+	return int64(v) // includes 0 as a valid, enforced cap
+}
+
+func (m *manager) AvailableBytes() uint64 {
+	c := m.CapacityBytes()
+	if c == 0 {
+		return 0
+	}
+	u := m.UsedBytes()
+	if u >= c {
+		return 0
+	}
+	return c - u
+}
+func (m *manager) AvailableCount() int64 {
+	c := m.CapacityCount()
+	if c == 0 {
+		return 0
+	}
+	u := m.UsedCount()
+	if u >= c {
+		return 0
+	}
+	return c - u
+}
+
+// getCacheUsage returns the CacheUsage for a given cacheID, creating one if it doesn't exist
+func (m *manager) getCacheUsage(cacheID string) *Usage {
+	if value, ok := m.cacheUsage.Load(cacheID); ok {
+		return value.(*Usage)
+	}
+
+	// Create new cache usage entry
+	newUsage := &Usage{}
+	actual, _ := m.cacheUsage.LoadOrStore(cacheID, newUsage)
+	return actual.(*Usage)
+}
+
+// getActiveCacheCount returns the current count of active caches
+func (m *manager) getActiveCacheCount() int64 {
+	return atomic.LoadInt64(&m.activeCacheCount)
+}
+
+// updateActiveCacheCount updates the active cache count when a cache transitions between active/inactive
+func (m *manager) updateActiveCacheCount(cacheUsage *Usage, wasActive bool) {
+	cacheUsage.mu.Lock()
+	isActive := cacheUsage.usedBytes > 0 || cacheUsage.usedCount > 0
+	cacheUsage.mu.Unlock()
+
+	if !wasActive && isActive {
```

Comment:
I don't think we got to discuss the active vs inactive state the various caches go through. What is the number of caches used to drive? I am guessing it's used in the fairness calculations?